daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -2094,6 +2107,59 @@ private[lf] object SBuiltin {
 
   }
 
+  private[speedy] final case class SBUpdateContractCache(
+      disclosures: ImmArray[DisclosedContract]
+  ) extends OnLedgerBuiltin(1) {
+
+    override protected def executeWithLedger(
+        args: util.ArrayList[SValue],
+        machine: Machine,
+        onLedger: OnLedger,
+    ): Control = {
+      checkToken(args, 0)
+
+      for (disclosedContract <- disclosures) {
+        val templateId = disclosedContract.templateId
+
+        if (machine.compiledPackages.pkgInterface.lookupTemplate(templateId).isLeft) {
+          crash(s"Template $templateId does not exist and it should")
+        }
+
+        val contractId = disclosedContract.contractId
+        val contract = SEValue(disclosedContract.argument)
+        val cachedContractSExpr =
+          if (machine.compiledPackages.pkgInterface.lookupTemplateKey(templateId).isRight) {
+            SELet1(
+              SEApp(SEVal(ContractKeyWithMaintainersDefRef(templateId)), Array(contract)),
+              SEApp(
+                SEVal(ToCachedContractDefRef(templateId)),
+                Array(contract, SEValue(SOptional(Some(SELocS(0).lookupValue(machine))))),
+              ),
+            )
+          } else {
+            SEApp(
+              SEVal(ToCachedContractDefRef(templateId)),
+              Array(contract, SEValue.None),
+            )
+          }
+        val cachedContract = Machine
```

Comment:
Question: is it reasonable to perform a Speedy pure eval per disclosed contract, or should we build a more complex expression and exec just the once?